### DISCUSSION TITLE
logout.cpl

### DISCUSSION THREAD

**Discussion by SirNibble** [prirun/p50em#9] [Q&A]
Wasn't there a compilmentary 'logout.cpl' to the 'login.cpl' ? 

I'm sure we had one at work, but going through the revisions of the pre-built systems I'm guessing a) I'm wrong, or b) it must have been something custom we had?

Any ideas?

(Mainly running Rev20 at the moment.)

**Comment by drboone** [comment]
There are two types of login and logout programs - the "external" ones in CMDNC0, and personal ones in your own UFD.  I don't _think_ there's a logout in the personal class, just in the "externals" class.  Any of these can be COMI files, CPL scripts, or compiled binaries (.SAVE or .RUN).

**Comment by SirNibble** [comment]
Hmm. I'm leaning toward me having a bout a false memory syndrome in that case. 

The late 80s was a long time ago ... :)

Thanks!

**Comment by SirNibble** [comment]
After more thought and trawling through "Appendix A External Login and Logout Programs" in the "Sys Admin Guide Rev 20.2 DOC5037-4LA 1986" I'm back to thinking we had 'custom' LOGIN & LOGOUT programs after all.

From that appendix ... (excuse any PDF OCR mistakes)

_Since Rev. 19, you can write separate external programs to monitor and control logins and logouts.
(Prior to Rev. 19, one program had to serve loth needs.)

At login time, the PRIMOS operating system looks for a program in
CMDNOO named LOGIN and runs it if it exists. At logout time, PRIMOS
looks first for a program named LOGOUT in CMDNOO. If the LOGOUT
program does not exist, FRIMOS then looks for LOGIN. (Suffixes are not
allowed on either name.)

These programs cannot be EPFs; they must be static-mode programs
created by SEG or LOAD._

Our login procedure would show you a system name banner, display a 'message of the day', list your latest emails, and then hand over to <USERMFD>USERXX>LOGIN.CPL as usual

I definitely remember now having a LOGOUT.CPL that did such things as saving the ECL command list and flushing a TEMP directory. Also, I remember our VDUs issued a 'clear screen' on logout. 

This leads me to believe that we must also have had a custom LOGOUT in CMDNC0

(Much of this was above my paygrade at the time. I was a tape and disk-pack monkey, and FSE baby-sitter, for most of my PR1ME admin/ops duties. I only really started fully being involved in customising systems as we moved into the era of the early EXL unix machines and Netware servers.)

So - my challenge to myself now is to dust off the Pascal knowledge (or PLP for the hell of it!) and see if I can recreate this functionality on my Rev20 machine. 

At least I ccan take a snap-shot of the machine on the host Pi4 in case I really bollix it up! ;)

**Comment by drboone** [comment]
I've got examples of our externals in FTN, and the seg split-load scripts and such, if you want such cheats.  We had some captive user accounts for which we'd disable the break key at login, and delete the "online" records for their sessions from the MIDAS files at logout.  I'd forgotten the CMDNC0 externals had to be static linked.

The personal ones could be nearly anything.  (And mine was LOGIN.RUN, and occasionally someone would edit it as a joke, stuff some CP0L in there, and save it.  Sigh.  OTOH, having `STAT US` list things like BITCH_SERVICE was somewhat amusing.)  I'll have to test for a personal LOGOUT.

Pascal has been a long time, but if you decide on PLP, I've done some of that more recently.

**Comment by drboone** [comment]
The aforementioned externals are here, in MAGSAV form:

https://sysovl.info/pages/blobs/prime/misc/boris_externals.tap

**Comment by SirNibble** [comment]
Ta!

I'll have a gander throught the BORIS tape tomorrow.

**Comment by SirNibble** [comment]
Putting the idea of the user LOGOUT.XXX aside for a second, I just want to check my logic for 'order of operation'.

At the very least, is this correct ... (ECL stuff invoked from my LOGIN.CPL

_Welcome to the Prime Computer 50-series emulator, running Primos rev 20.2.8


Login please.
login
User id? chris
Password?

CHRIS (user 2) logged in Tuesday, 26 Jan 21 00:33:12.
Welcome to PRIMOS version 20.2.8
Copyright (c) Prime Computer, Inc. 1985.
Last login Tuesday, 26 Jan 21 00:26:40.

++++    RESULT OF CMDNC0>LOGIN HERE    ++++ 
++++    RESULT OF CMDNC0>LOGIN HERE    ++++
++++    RESULT OF CMDNC0>LOGIN HERE    ++++

[EDIT_CMD_LINE Rev. 21.0.4 Copyright (c) 1988, Prime Computer, Inc.]
Editing command input with ECL. (EDIT_CMD_LINE)
Restored 200 history commands from `<USERS>CHRIS>.CMD_HISTORY'. (EDIT_CMD_LINE)

<USERS>CHRIS> 

  //session
  //session
  //session

<USERS>CHRIS>logout

++++    RESULT OF CMDNC0>LOGOUT HERE    ++++ 
++++    RESULT OF CMDNC0>LOGOUT HERE    ++++
++++    RESULT OF CMDNC0>LOGOUT HERE    ++++

CHRIS (user 2) logged out Tuesday, 26 Jan 21 00:33:36.
Time used: 00h 00m connect, 00m 00s CPU, 00m 00s I/O.
OK,_


That feels right from memory, but best to check.

And I guess I'm better off going with Prime's Pascal (Rev. 20.2.B2 in my case) rather than the Sheffield V2.7 I got working from your site?

**Comment by drboone** [comment]
The general order looks right to me.  And a personal LOGIN would run after the CMDNC0 one.

Seems like the Sheffield compiler was a bit old, but if you got it running, you've gone farther than I.

**Comment by drboone** [comment]
At Rev20, a personal LOGOUT.CPL doesn't seem to be invoked.

**Comment by SirNibble** [comment]
A pair of very quick & dirty Sheffield Pascal programs that purely read CMDNC0>BANNER_IN and _OUT files shows a slight hiccup in the running order. The logout procedure is PRIMOS Logout -> CMDNC0>LOGOUT

CHRIS (user 4) logged in Tuesday, 26 Jan 21 14:31:36.
Welcome to PRIMOS version 20.2.8
Copyright (c) Prime Computer, Inc. 1985.
Last login Tuesday, 26 Jan 21 14:28:24.
(1 other user under this id.)

*****************************************************************************

  ++++ BANNER_IN TEXT ++++

*****************************************************************************

[EDIT_CMD_LINE Rev. 21.0.4 Copyright (c) 1988, Prime Computer, Inc.]
Editing command input with ECL. (EDIT_CMD_LINE)
Restored 200 history commands from `<USERS>CHRIS>.CMD_HISTORY'. (EDIT_CMD_LINE)

<USERS>CHRIS>lo
Saved 200 history commands to `<USERS>CHRIS>.CMD_HISTORY'. (EDIT_CMD_LINE)

CHRIS (user 4) logged out Tuesday, 26 Jan 21 14:31:40.
Time used: 00h 00m connect, 00m 01s CPU, 00m 00s I/O.
*****************************************************************************

            ++++ BANNER_OUT TEXT ++++

*****************************************************************************
OK,
Prime session disconnected
Connection closed by foreign host.


Can't really investigate more today, but sorta fun so far!

(btw: Gone with Sheffield as having problems with Prime Pascal. Compiling & seg loading OK, but throws up LINKAGE faults at run time. That can wait for now.)

**Comment by drboone** [comment]
You need to share the PASCAL library, `CO SYSTEM>PASCAL.SHARE.COMI`.  The example rev20 system has this broken too, I see.

**Reply by SirNibble** [reply]
> You need to share the PASCAL library, `CO SYSTEM>PASCAL.SHARE.COMI`. The example rev20 system has this broken too, I see.

Aye, done that in PRIMOS.COMI - but I'd not put PASCAL_LIBRARY.RUN in the LIBRARIES* directory. At least regular 'hello world' stuff now works.

IIRC I used the excellent INVENTORY* to patch-work install bits from the R19, R20 and R21 examples until the various Shares and Search Rules were happy. It was a bit scattergun, and I wish I'd made proper notes now!

**Comment by drboone** [comment]
Would be nice to have more complete / orthogonal product sets!

You wouldn't ordinarily need both the static-mode shares _and_ the EPF library.  You _would_ need the compiler and its library to match.

**Comment by drboone** [comment]
No, sorry, that's not necessarily right.  If the compiler is a .save, then you'd need the shares.

**Reply by SirNibble** [reply]
> No, sorry, that's not necessarily right. If the compiler is a .save, then you'd need the shares.

Oh, sorry, I wasn't clear enough.  The SHARES are there (see output from INVENTORY* util)

 Pascal (8522)
 *************
 <SYS202>CMDNC0>PASCAL.SAVE                                                    
 <SYS202>SYSTEM>PA2064
 <SYS202>SYSTEM>PA2065
 <SYS202>SYSTEM>PA2066
 <SYS202>SYSTEM>PASCAL.SHARE.COMI
 <SYS202>LIB>PASLIB.BIN
 <SYS202>SYSOVL>PASCALDATA
 <SYS202>SYSCOM>ERRD.INS.PASCAL
 <SYS202>SYSCOM>KEYS.INS.PASCAL
 <SYS202>SYSCOM>ERRD.PASCAL
 <SYS202>SYSCOM>KEYS.PASCAL

as well as ...

<SYS202>CMDNC0>PASCAL.RUN                        (On the R20 example)
<SYS202>LIBRARIES*>PASCAL_LIBRARY.RUN      (From R21? Can't remember)

But I think my 'patch-work' install has complicated things with the Rev20 PASCAL.RUN  - lord knows when/how I moved the PASCAL_LIBRARY.RUN (from the R21 machine? Can't remember) file to CDMCN0, but putting it back in LIBRARIES* stopped runtime link errors from Pascal Rev20 compiled EPFs and SEGs.

**Comment by SirNibble** [comment]
Anyway - looks like my next step at the weekend is to troll through the Subroutines Reference Guide to remind myself how CP$ (?? that sounds right??) works to call the CPL.

It's probably a good job my printer is out-of-service at the moment as I think I'd be surrounded by a small forest-worth of re-printed manuals by now ... ;)

**Comment by drboone** [comment]
I suspect there's an (undocumented) way to directly call ECL to save its history, but I haven't worked it out yet.  I have a .MAP file for ECL$LIB.RUN, but it's compressed with a custom tool, so I need to find the decompressor.

CP$ is the right "invoke a command" subroutine.

**Comment by SirNibble** [comment]
Small, silly, victories are the best ..

So this is very bare bones with no sanity checking
```
program main(input,output);

const
  %INCLUDE 'SYSCOM>ERRD.INS.PASCAL';
  %INCLUDE 'SYSCOM>KEYS.INS.PASCAL';
  %INCLUDE 'SYSCOM>A$KEYS.INS.PASCAL';

var
  status, code : integer;
  cmd : string;

  procedure cp$(var ccmd:string; stat:integer; rc:integer);extern;

begin

  writeln('Ready to engage CPL ...');
  cmd := 'r <users>sman>goodbye.cpl';
  cp$(cmd,status,code);
  writeln(status);
  writeln(code);
  writeln('And done ..');
  writeln;
end.
```

goodbye.cpl is just 

`type 'Goodbye from CPL'
`
for now

And so we get ...

```
<USERS>SMAN>lo
Saved 139 history commands to `<USERS>SMAN>.CMD_HISTORY'.

SMAN (user 2) logged out Saturday, 06 Feb 21 12:36:16.
Time used: 00h 01m connect, 00m 01s CPU, 00m 00s I/O.
Ready to engage CPL ...
Goodbye from CPL
         0
         0
And done ..


Prime session disconnected

```


Which confirms that calling another program on LOGOUT works.

Will have to fettle what actually works within the CPL, but happy the basic theory works.

Did throw me for a while where I could get the program to BIND & RUN but not complete a SEG LOAD until I remembered that it had to be program main(input,output); rather than program yadayada(input,output);

Fun!

**Comment by hashbackup** [comment]
Can your command be "r goodbye.cpl" to make it general for any user,
or has Primos already disassociated your home directory by the time
logout runs?

**Comment by drboone** [comment]
Also, if it hasn't disassociated already, "home" may not currently be "origin", so you might want to call `at$or(k$seth, code);` before invoking the CPL.

**Reply by drboone** [reply]
Just wandered through the 19.2 source.  PRIMOS cleans up HOME and CURRENT, but not ORIGIN, before running external logout.  (KS>LO_FATAL.PLP)

**Comment by SirNibble** [comment]
That was to be my next step, using the suggested at$or(k$seth, code), so it should work for any user.

And I suppose I really should do some checking for the file existing.

**Comment by SirNibble** [comment]
Good Lord, but my Pascal is rusty. Spent far too long banging my head against a 'bad variable' compile error as I'd got my procedure definitions wrong.

But, yep, using the at$or sub gets you home to run the CPL. (Yes ... I must put error checking in!)

```
program main(input,output);

const
  %INCLUDE 'SYSCOM>ERRD.INS.PASCAL';
  %INCLUDE 'SYSCOM>KEYS.INS.PASCAL';
  %INCLUDE 'SYSCOM>A$KEYS.INS.PASCAL';

var
  status, code : integer;

  procedure at$or(key:integer;  var rc:integer);extern;
  procedure cp$(cmd:string; var stat:integer; var rc:integer);extern;

begin
  writeln('Going to OR ...');
  at$or(k$seth,code);
  writeln('R LOGOUT.CPL');
  cp$('R LOGOUT.CPL',status,code);
  write(status);writeln(code);
  writeln('And done ..');
  writeln;
end.

```

Gives ...
```

<USERS>SMAN>or
<USERS>SMAN>ld

<USERS>SMAN (ALL access)
6 records in this directory, 324 total records out of quota of 0.

5 Files.

.ABBREV          .CMD_HISTORY     LOGIN.CPL        LOGOUT.CPL
WASTE.BAS

8 Directories.

CC               FOO              INFOTEST         LI
MY_LOGIN         PAS              PLP              UN02

<USERS>SMAN>slist logout.cpl
type 'Goodbye from PiPrime01'
date

<USERS>SMAN>a <misc>guest21
<USERS>SMAN>ld

<MISC>GUEST21 (ALL access)
90 records in this directory, 5005 total records out of quota of 0.

5 Files.

DOWN.CPL         KERMIT.RUN       LISTING.CPL      LOGIN.CPL
UP.CPL

5 Segment Directories.

D_&SAVEDLISTS&   D_VOC            D_VOCLIB         VOC
VOCLIB

7 Directories.

&SAVEDLISTS&     ANDREW           DENNIS           INFORMATION
PEN_TEST         PHIL             TEST

<USERS>SMAN>lo
Saved 158 history commands to `<USERS>SMAN>.CMD_HISTORY'.

SMAN (user 3) logged out Sunday, 07 Feb 21 00:14:24.
Time used: 00h 02m connect, 00m 02s CPU, 00m 00s I/O.
Going to OR ...
R LOGOUT.CPL
Goodbye from PiPrime01
07 Feb 21 00:14:24 Sunday
         0         0
And done ..


Prime session disconnected

```

**Comment by drboone** [comment]
Excellent!
